--- Ethernet.cpp	2013-05-17 14:22:15.000000000 -0400
+++ Ethernet.cpp	2014-08-26 11:14:00.083713733 -0400
@@ -7,13 +7,12 @@
   0, 0, 0, 0 };
 uint16_t EthernetClass::_server_port[MAX_SOCK_NUM] = { 
   0, 0, 0, 0 };
+static DhcpClass s_dhcp;
 
 int EthernetClass::begin(uint8_t *mac_address)
 {
-  static DhcpClass s_dhcp;
   _dhcp = &s_dhcp;
 
-
   // Initialise the basic info
   W5100.init();
   W5100.setMACAddress(mac_address);
