--- /usr/share/arduino/libraries/Ethernet/utility/socket.cpp	2014-06-13 15:35:45.257740458 -0700
+++ /oc/usr/share/arduino/libraries/Ethernet/utility/socket.cpp	2014-08-10 17:20:30.273142691 -0700
@@ -248,6 +248,7 @@
 {
   uint8_t head[8];
   uint16_t data_len=0;
+  uint16_t data_copied=0;
   uint16_t ptr=0;
 
   if ( len > 0 )
@@ -268,10 +269,6 @@
       data_len = head[6];
       data_len = (data_len << 8) + head[7];
 
-      W5100.read_data(s, (uint8_t *)ptr, buf, data_len); // data copy.
-      ptr += data_len;
-
-      W5100.writeSnRX_RD(s, ptr);
       break;
 
     case SnMR::IPRAW :
@@ -285,10 +282,6 @@
       data_len = head[4];
       data_len = (data_len << 8) + head[5];
 
-      W5100.read_data(s, (uint8_t *)ptr, buf, data_len); // data copy.
-      ptr += data_len;
-
-      W5100.writeSnRX_RD(s, ptr);
       break;
 
     case SnMR::MACRAW:
@@ -297,17 +290,22 @@
       data_len = head[0];
       data_len = (data_len<<8) + head[1] - 2;
 
-      W5100.read_data(s,(uint8_t*) ptr,buf,data_len);
-      ptr += data_len;
-      W5100.writeSnRX_RD(s, ptr);
       break;
 
     default :
       break;
     }
+
+    if (data_len > 0)
+    {
+        data_copied = (data_len <= len) ? data_len : len;// only copy the len bytes, rest of is discarded.
+        W5100.read_data(s, (uint8_t *)ptr, buf, data_copied); // data copy.
+        ptr += data_len;
+        W5100.writeSnRX_RD(s, ptr);
+    }
     W5100.execCmdSn(s, Sock_RECV);
   }
-  return data_len;
+  return data_copied;
 }
 
 
